File: 021_students/students.py
#!/usr/bin/env python

import sys

students = {} 
for line in sys.stdin:

    id, score = line.rstrip().split('-')
    if id not in students or students[id] < score:
        students[id] = score

for id in sorted(students, reverse=True):
    print "%s %s" % (id, students[id])

File: 021_students/students.rb
#!/usr/bin/env ruby

students = {}
ARGF.each do |str|

    id, score = str.split('-').map(&:to_i)

    if not students.has_key?(id) or students[id] < score
        students[id] = score
    end
end

students.sort_by {|k, v| v}.reverse.each do |k, v|
    puts "#{k}-#{v}"
end
